lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -622,20 +622,33 @@ def add_mpi_data_for_network_tokenization_card(post, payment, options)
 
         post[:mpiData] = {}
         post[:mpiData][:authenticationResponse] = 'Y'
-        post[:mpiData][:cavv] = payment.payment_cryptogram
+        if NETWORK_TOKENIZATION_CARD_SOURCE[payment.source.to_s].nil? && options[:switch_cryptogram_mapping_nt]
+          post[:mpiData][:tokenAuthenticationVerificationValue] = payment.payment_cryptogram
+        else
+          post[:mpiData][:cavv] = payment.payment_cryptogram
+        end
         post[:mpiData][:directoryResponse] = 'Y'
         post[:mpiData][:eci] = payment.eci || '07'
       end
 
-      def add_recurring_contract(post, options = {})
-        return unless options[:recurring_contract_type]
+      def add_recurring_contract(post, options = {}, payment = nil)
+        return unless options[:recurring_contract_type] || (payment.is_a?(NetworkTokenizationCreditCard) && options[:switch_cryptogram_mapping_nt])
 
-        post[:recurring] = {}
-        post[:recurring][:contract] = options[:recurring_contract_type]
-        post[:recurring][:recurringDetailName] = options[:recurring_detail_name] if options[:recurring_detail_name]
-        post[:recurring][:recurringExpiry] = options[:recurring_expiry] if options[:recurring_expiry]
-        post[:recurring][:recurringFrequency] = options[:recurring_frequency] if options[:recurring_frequency]
-        post[:recurring][:tokenService] = options[:token_service] if options[:token_service]
+        post[:recurring] ||= {}
+
+        if options[:recurring_contract_type]
+          post[:recurring][:contract] ||= options[:recurring_contract_type]
+          post[:recurring][:recurringDetailName] = options[:recurring_detail_name]
+          post[:recurring][:recurringExpiry] = options[:recurring_expiry]
+          post[:recurring][:recurringFrequency] = options[:recurring_frequency]
+          post[:recurring][:tokenService] ||= options[:token_service]
+        elsif payment.is_a?(NetworkTokenizationCreditCard) && options[:switch_cryptogram_mapping_nt]
```

Comment:
In this new `elsif` scenario we aren't setting the `recurringDetailName` `recurringExpiry` or `recurringFrequency` Is this intentional? If so why do we not want to set them in that case?

Reply:
Also I still don't know if these fields are applicable to apple_pay and google_pay since those aren't automatically recurring payments. I think we should only set them automatically for true NetworkTokens.

Reply:
I agree that this should only be for NetworkTokens and that the NetworkToken check should be outside the if statment

```
return unless options[:recurring_contract_type] || (payment.try(:source) == :network_token && options[:switch_cryptogram_mapping_nt])

        post[:recurring] ||= {}

        if options[:recurring_contract_type]
          post[:recurring][:contract] ||= options[:recurring_contract_type]
          post[:recurring][:recurringDetailName] = options[:recurring_detail_name]
          post[:recurring][:recurringExpiry] = options[:recurring_expiry]
          post[:recurring][:recurringFrequency] = options[:recurring_frequency]
          post[:recurring][:tokenService] ||= options[:token_service]
        end

if payment.try(:source) == :network_token && options[:switch_cryptogram_mapping_nt]
          post[:recurring][:contract] = 'EXTERNAL'
          post[:recurring][:tokenService] = case payment.brand
                                            when 'visa' then 'VISATOKENSERVICE'
                                            else 'MCTOKENSERVICE'
                                            end
        end
```

Reply:
I missed the other recurring fields in the previous approach trying to avoid the `recurring=>{nil}` ocurrence.
I think the current one solves that.